handler/src/test/java/io/netty/handler/ssl/JdkSslEngineTest.java

```diff
@@ -22,110 +22,134 @@
+    private static final String FALLBACK_APPLICATION_LEVEL_PROTOCOL = "my-protocol-http1_1";
+    private static final String APPLICATION_LEVEL_PROTOCOL_NOT_COMPATIBLE = "my-protocol-FOO";
+
+    @Parameterized.Parameters(name = "{index}: providerType = {0}, bufferType = {1}")
+    public static Collection<Object[]> data() {
+        List<Object[]> params = new ArrayList<Object[]>();
+        for (ProviderType providerType : ProviderType.values()) {
+            for (BufferType bufferType : BufferType.values()) {
+                params.add(new Object[]{providerType, bufferType});
             }
-            ApplicationProtocolConfig clientApn = failingNegotiator(Protocol.NPN,
-                    PREFERRED_APPLICATION_LEVEL_PROTOCOL);
-            ApplicationProtocolConfig serverApn = acceptingNegotiator(Protocol.NPN,
-                    APPLICATION_LEVEL_PROTOCOL_NOT_COMPATIBLE);
-            setupHandlers(serverApn, clientApn);
-            assertTrue(clientLatch.await(2, TimeUnit.SECONDS));
-            assertTrue(clientException instanceof SSLHandshakeException);
-        } catch (SkipTestException e) {
-            // NPN availability is dependent on the java version. If NPN is not available because of
-            // java version incompatibility don't fail the test, but instead just skip the test
-            assumeNoException(e);
         }
+        return params;
+    }
+
+    private final ProviderType providerType;
+
+    private Provider provider;
+
+    public JdkSslEngineTest(ProviderType providerType, BufferType bufferType) {
+        super(bufferType);
+        this.providerType = providerType;
     }
 
     @Test
-    public void testNpnNoCompatibleProtocolsServerHandshakeFailure() throws Exception {
+    public void testTlsExtension() throws Exception {
         try {
-            // Typical code will not have to check this, but will get a initialization error on class load.
-            // Check in this test just in case we have multiple tests that just the class and we already ignored the
-            // initialization error.
-            if (!JdkNpnSslEngine.isAvailable()) {
-                throw tlsExtensionNotFound(Protocol.NPN);
-            }
-            ApplicationProtocolConfig clientApn = acceptingNegotiator(Protocol.NPN,
-                    PREFERRED_APPLICATION_LEVEL_PROTOCOL);
-            ApplicationProtocolConfig serverApn = failingNegotiator(Protocol.NPN,
-                    APPLICATION_LEVEL_PROTOCOL_NOT_COMPATIBLE);
-            setupHandlers(serverApn, clientApn);
-            assertTrue(serverLatch.await(2, TimeUnit.SECONDS));
-            assertTrue(serverException instanceof SSLHandshakeException);
+            providerType.activate(this);
+            ApplicationProtocolConfig apn = failingNegotiator(providerType.protocol(),
+                PREFERRED_APPLICATION_LEVEL_PROTOCOL);
+            setupHandlers(apn);
+            runTest();
```

Comment:
@Scottmitch AFAIK the test is a unchanged from what it was before. Am I missing something?